src/routing/effects/index.ts

```diff
@@ -0,0 +1,372 @@
+};
+
+const crawl = (index, parent, parentObject, states, routesMap) => {
+  const obj = !isEmpty(parentObject) ? parentObject : defaultState;
+
+  return Object.keys(states).map((key) => {
+    const state = states[key];
+    const data = state.$;
+    const local = urlBuilder.compile(data.url, {
+      strict: false,
+      caseInsensitive: false,
+    });
+
+    const qualified = parent ? parent.append(local) : local;
+
+    const newObj = merge(obj, {
+      state,
+      components: data.component ? obj.components.concat(data.component) : obj.components,
+      requestData:
+        obj.requestData.concat([{
+          remoteData: data.remoteData || null,
+          conditions: data.conditions || null,
+          redirect: data.redirect || null,
+          data: data.data || {}
+        }]),
+      conditions: obj.conditions.concat([data.conditions || null]),
+      data: merge(obj.data, data.data),
+      redirect: data.redirect,
+      remoteData: obj.remoteData.concat([[data.remoteData || null]]),
+      componentsWithNeeds:
+        obj.componentsWithNeeds.concat([{
+          component: data.component,
+          policy: 'Loaded',
+          needs: Object.keys(data.remoteData || {})
+        }])
+    });
+
+    const current = index.concat([[newObj, qualified]]);
+
+    routesMap.set(state, { data: newObj, urlMatcher: qualified });
+
+    return Object.keys(state).length === 0 ? current : current.concat(...crawl(
+      index,
+      qualified,
+      newObj,
+      omit(['$'], state),
+      routesMap,
+    ));
+  });
+};
+
+const matchUrl = (routes, pathname, query): State => pathname && pathname !== '/' ?
+  Maybe.of(routes.find(([{ }, matcher]) => !isNil(matcher.exec(pathname, query))))
+    .map(([data, matcher]) => ({ data, params: matcher.exec(pathname, query), pathname }))
+    .defaultTo({ data: {}, params: {}, pathname })
+    : Maybe.of(routes.find(([data, matcher]) => data.data.defaultRoute))
+    .map(([data, matcher]) => ({ data, params: matcher.exec(pathname), pathname }))
+    .defaultTo({ data: {}, params: {}, pathname });
+
+const getStateKey = (state, routesMap) => routesMap.get(state) || routesMap.get(Array.from(routesMap.keys())
+.find(key => equals(state, JSON.parse(JSON.stringify(key)))));
+
+const changeState = ({ data, matcher, params = {}, historyChange = 'push', location }) => {
+  const newPath = matcher.format(params);
+
+  const newUrl = [window.location.origin, newPath.replace(/^\//, '')].join('/');
+  const replace = data.data.replace || historyChange === 'replace';
+
+  if (historyChange !== 'back' && !equals(location.path, newPath)) {
+    history[replace ? 'replaceState' : 'pushState']({ state: JSON.stringify(data.state), params }, '', newUrl);
+
+    window.scrollTo(0, 0);
+  }
+
+  return {
+    data,
+    params: merge(location.query, params),
+    pathname: newPath,
+  };
+};
+
+const findState = ({ route: state, params, data: { replace, back } }, routes, location, routesMap): any =>
```

Comment:
It'd actually be cool to see this return a `Maybe` and not the value of a `Maybe` :alyson: